controller/service.go

```diff
@@ -180,3 +185,132 @@ func (c *controller) allocateIP(key string, svc *v1.Service) (net.IP, error) {
+	// start converging from a clean slate.
+	if len(svc.Status.LoadBalancer.Ingress) > 1 {
+		lbIP = net.ParseIP(svc.Status.LoadBalancer.Ingress[0].IP)
+		lbIP2 = net.ParseIP(svc.Status.LoadBalancer.Ingress[1].IP)
+	}
+
+	// It's possible the config mutated and the IP we have no longer
+	// makes sense. If so, clear it out and give the rest of the logic
+	// a chance to allocate again.
+	if lbIP != nil && lbIP2 != nil {
+		// This assign is idempotent if the config is consistent,
+		// otherwise it'll fail and tell us why.
+		if err := c.ips.AssignDual(key, lbIP, lbIP2, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
+			l.Log("event", "clearAssignment", "reason", "notAllowedByConfig", "msg", "current IP not allowed by config, clearing")
+			c.clearServiceState(key, svc)
+			lbIP = nil
+		}
+
+		if lbIP != nil {
+			// The user might also have changed the pool annotation, and
+			// requested a different pool than the one that is currently
+			// allocated.
+			desiredPool := svc.Annotations["metallb.universe.tf/address-pool"]
+			if desiredPool != "" && c.ips.Pool(key) != desiredPool {
+				l.Log("event", "clearAssignment", "reason", "differentPoolRequested", "msg", "user requested a different pool than the one currently assigned")
+				c.clearServiceState(key, svc)
+				lbIP = nil
+			}
+		}
+	} else {
+		c.clearServiceState(key, svc)
+		lbIP = nil
+	}
+
+	// The (singular) svc.Spec.LoadBalancerIP is ignored for dual-stack
+	if svc.Spec.LoadBalancerIP != "" {
+		l.Log("event", "loadBalancerIP", "reason", "N/A", "msg", "loadBalancerIP ignored for dual-stack")
+	}
+
+	if requestedIPs := svc.Annotations["metallb.universe.tf/load-balancer-ips"]; requestedIPs != "" {
+		// Until a svc.Spec.LoadBalancerIPs exists we use an annotation.
+		// requestedIPs must be a comma-separated list of 2 addresses, one from each family.
+		ips := strings.Split(requestedIPs, ",")
+		if len(ips) != 2 {
+			l.Log("op", "allocateIP", "load-balancer-ips", len(ips), "msg", "Must be two addresses")
+			return true
+		}
+		if lbIP = net.ParseIP(strings.TrimSpace(ips[0])); lbIP == nil {
+			l.Log("op", "allocateIP", "load-balancer-ips", ips[0], "msg", "Invalid addresses")
+			return true
+		}
+		if lbIP2 = net.ParseIP(strings.TrimSpace(ips[1])); lbIP2 == nil {
+			l.Log("op", "allocateIP", "load-balancer-ips", ips[1], "msg", "Invalid addresses")
+			return true
+		}
+		if (lbIP.To4() == nil) == (lbIP2.To4() == nil) {
+			l.Log("op", "allocateIP", "load-balancer-ips", requestedIPs, "msg", "Same family")
+		}
+
+		// Try to assign the requested IPs
+		if err := c.ips.AssignDual(key, lbIP, lbIP2, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
+			l.Log("op", "allocateIP", "error", err, "msg", "Can't assign requested IPs")
+			return true
+		}
+	}
+
+	// If lbIP's is still nil at this point, try to allocate.
+	if lbIP == nil {
+		if !c.synced {
```

Comment:
this is probably taken from the other one, but I wonder why it can't be done from the beginning or even before invoking `SetBalancer`

Reply:
I am unsure, so I keept Dave Andersons code rather than risking to mess something up.